yellow-paper/docs/gas-and-fees/gas-metering-and-fees.md

```diff
@@ -0,0 +1,683 @@
+    public_kernel_public_inputs.end_non_revertible.gas_used.l2_gas_used
+    + public_kernel_public_inputs.end.gas_used.l2_gas_used
+    # recall we have already included teardown gas in the non-revertible phase
+
+kernel_public_inputs.end.gas_used.l1_gas_used :=
+    public_kernel_public_inputs.end_non_revertible.gas_used.l1_gas_used
+    + public_kernel_public_inputs.end.gas_used.l1_gas_used
+    # recall we have already included teardown gas in the non-revertible phase
+
+fee_per_da_gas := public_kernel_public_inputs.constants.pending_header.global_variables.fee_per_da_gas
+fee_per_l2_gas := public_kernel_public_inputs.constants.pending_header.global_variables.fee_per_l2_gas
+fee_per_l1_gas := public_kernel_public_inputs.constants.pending_header.global_variables.fee_per_l1_gas
+
+kernel_public_inputs.transaction_fee :=
+    fee_per_da_gas * kernel_public_inputs.end.gas_used.da_gas_used
+    + fee_per_l2_gas * kernel_public_inputs.end.gas_used.l2_gas_used
+    + fee_per_l1_gas * kernel_public_inputs.end.gas_used.l1_gas_used
+    + kernel_public_inputs.gas_limits.inclusion_fee
+```
+
+### Teardown
+
+As noted above, the user specifies `teardown_da_gas`, `teardown_l2_gas`, and `teardown_l1_gas`, which are the amounts of DA, L2, and L1 gas that the user has agreed to pay for the teardown function.
+
+I.e., teardown is not metered for the purpose of calculating `_gas_used`, and instead the public VM is is provided with `*GasLeft` initialized to `teardown_*_gas`, but the final values of `*GasLeft` are ignored.
+
+The reasons for this are:
+- we want to know the exact transaction fee **prior** to teardown execution so it can be set in the public context
+- we want to give an assurance to the sequencer (and any full node) that the teardown function has finite gas limits
+
+After executing the teardown function, the teardown kernel will assert that the `transaction_fee` that was made available in the public context is equal to `kernel_public_inputs.transaction_fee`.
+
+This ensures that the `pay_fee` function on the enshrined gas token contract **can** be called with the correct amount.
+
+### Validating fee payment
+
+During execution of the teardown function, the public VM will need to inform the kernel **if** the `pay_fee` function was called.
+
+This is done via a `feePaid` boolean in the MachineState.
+
+This field is included in the `PublicCircuitPublicInputs` struct as `fee_paid`.
+
+The public kernel teardown will then set:
+
+```
+kernel_public_inputs.transaction_fee :=
+    public_inputs.fee_paid ? kernel_public_inputs.transaction_fee : 0
+```
+
+:::warning Questions
+Is there a more clean way to ensure the fee is paid other than this very specific boolean?
+How does the VM know that the `pay_fee` function has been called?
+Would we rather the VM return function calls that were made, and have the kernel check for the `pay_fee` function?
+How do we convince the kernel that the machine state provided in the public inputs is correct?
```

Comment:
I think we should come up with a way of enshrining `pay_fee`. Say, in the base rollup for example. All it does is subtract an amount of the gas token from a balance. I think it would significantly simplify things.

If we did this in the base rollup then it can be used for paying for private only transactions and we wouldn't need to prove a VM and kernel iteration for it.

Does this sound sensible to you @iAmMichaelConnor